### PR TITLE
Helm release

### PR DESCRIPTION
**Pull request by mkyc**
## Description

This PR: 
 - is merging prereqs and operator helm charts into one eunomia-operator chart
 - is introducing publication of helm charts to GitHub Pages in gh-pages branch
 - obsoletes PR https://github.com/KohlsTechnology/eunomia/pull/85

## Type of change

* New feature: publish helm charts after creation of tag
* Breaking change: merge of prereqs and operator helm charts
* This change requires a documentation update: I'm unable to fix examples/cluster documentation as it's whole big chunk of reality and requires separate work

### CODE REVIEW

File: scripts/build/push-to-pages.sh [hunk 1]
```diff
@@ -1,7 +1,6 @@
 git -C $GITHUB_PAGES_DIR config user.email "travis@users.noreply.github.com"
 git -C $GITHUB_PAGES_DIR config user.name travis
-# git -C $GITHUB_PAGES_DIR config credential.helper 'store --file ~/.travis.git.credentials'
-# echo https://$GIT_USERNAME:$GIT_PASSWORD@github.com >> ~/.travis.git.credentials
+git -C $GITHUB_PAGES_DIR config core.sshCommand 'ssh -i github_deploy_key'
```
Comment (akavel) [inline scripts/build/push-to-pages.sh]: Could you please explain a bit more in the commit message on "why?" of this commit? I don't really understand what's going on here, what was bad in the previous code and why the new one is better :/ (also, similar for the change in build-chart-repo.sh)

File: scripts/build-local.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Comment (akavel) [inline scripts/build-local.sh]: Please add a short comment here, at least with the same info as in the commit message.

File: Makefile [hunk 3]
```diff
@@ -55,3 +61,10 @@ e2e-test-images: manager
 travis-deploy-images: manager
 	docker login -u ${DOCKER_USER} -p ${DOCKER_PASSWORD} ${REGISTRY}
 	./scripts/build-images.sh ${REPOSITORY} true
+
+publish-chart-repo:
+	./scripts/build/checkout-rebase-pages.sh
+	./scripts/build/build-chart-repo.sh
+	./scripts/build/push-to-pages.sh
+
+travis-release-deploy: travis-deploy-images	publish-chart-repo
```
Comment (akavel) [inline Makefile]: `travis-deploy-release`, for consistency with the naming of the other make targets? or maybe just `travis-release`?

File: .travis.yml [hunk 4]
```diff
@@ -11,16 +11,12 @@ env:
...
   - travis_retry go mod vendor
   - go get golang.org/x/lint/golint
-  - openssl aes-256-cbc -K $encrypted_18f270609d30_key -iv $encrypted_18f270609d30_iv -in github_deploy_key.enc -out github_deploy_key -d
-  - chmod 600 github_deploy_key
-  - eval $(ssh-agent -s)
-  - ssh-add github_deploy_key
+  - if ! [[ -z "${TRAVIS_TAG}" ]]; then openssl aes-256-cbc -K $encrypted_18f270609d30_key -iv $encrypted_18f270609d30_iv -in github_deploy_key.enc -out github_deploy_key -d; fi
+  - if ! [[ -z "${TRAVIS_TAG}" ]]; then chmod 600 github_deploy_key; fi
+  - if ! [[ -z "${TRAVIS_TAG}" ]]; then eval $(ssh-agent -s); fi
+  - if ! [[ -z "${TRAVIS_TAG}" ]]; then ssh-add github_deploy_key; fi
```
[the remark this answers is not in the document]
Comment (mkyc) [reply .travis.yml]: Done.

File: .travis.yml [hunk 5]
```diff
@@ -11,16 +11,12 @@ env:
...
-  - if [ "${TRAVIS_PULL_REQUEST}" = "false" ]; then chmod 600 github_deploy_key; fi
-  - if [ "${TRAVIS_PULL_REQUEST}" = "false" ]; then eval $(ssh-agent -s); fi
-  - if [ "${TRAVIS_PULL_REQUEST}" = "false" ]; then ssh-add github_deploy_key; fi  
   - travis_retry go mod vendor
   - go get golang.org/x/lint/golint
-  - openssl aes-256-cbc -K $encrypted_18f270609d30_key -iv $encrypted_18f270609d30_iv -in github_deploy_key.enc -out github_deploy_key -d
-  - chmod 600 github_deploy_key
-  - eval $(ssh-agent -s)
-  - ssh-add github_deploy_key
+  - if ! [[ -z "${TRAVIS_TAG}" ]]; then openssl aes-256-cbc -K $encrypted_18f270609d30_key -iv $encrypted_18f270609d30_iv -in github_deploy_key.enc -out github_deploy_key -d; fi
```
[the remark this answers is not in the document]
Comment (mkyc) [reply .travis.yml]: Done.

File: scripts/build/build-chart-repo.sh [hunk 6]
```diff
@@ -0,0 +1,13 @@
+echo '>> Building charts...'
```
[the remark this answers is not in the document]
Comment (mkyc) [reply scripts/build/build-chart-repo.sh]: v1 and v2 of this comment done. About "edit2" @seanmalloy  please comment what should be correct way.

File: scripts/build/build-chart-repo.sh [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+find "$HELM_CHARTS_SOURCE" -mindepth 1 -maxdepth 1 -type d | while read chart; do
+  version=${TRAVIS_TAG} envsubst < $chart/values.yaml.tpl > $chart/values.yaml
+  version=${TRAVIS_TAG} envsubst < $chart/Chart.yaml.tpl  > $chart/Chart.yaml
```
[the remark this answers is not in the document]
Comment (mkyc) [reply scripts/build/build-chart-repo.sh]: Done.

File: DEVELOPMENT.md [hunk 8]
```diff
@@ -46,8 +46,8 @@ The most efficient way to develop the operator locally is run the code on your l
 ```
 minikube start
 kubectl apply -f ./deploy/crds/eunomia_v1alpha1_gitopsconfig_crd.yaml
-export JOB_TEMPLATE=./deploy/helm/operator/eunomia-templates/job.yaml
-export CRONJOB_TEMPLATE=./deploy/helm/operator/eunomia-templates/cronjob.yaml
+export JOB_TEMPLATE=./deploy/helm/eunomia-operator/eunomia-templates/job.yaml
```
[the remark this answers is not in the document]
Comment (mkyc) [reply DEVELOPMENT.md]: That is good idea, but I'd like some confirmation from @seanmalloy that my general approach to merge those two charts is correct.

File: scripts/build/push-to-pages.sh (hunk 1 above)
Comment (mkyc) [reply scripts/build/push-to-pages.sh]: I updated commit message.

File: scripts/build-local.sh (hunk 2 above)
Comment (mkyc) [reply scripts/build-local.sh]: I renamed script to deploy-to-minikube.sh and added little comment.

File: Makefile (hunk 3 above)
Comment (mkyc) [reply Makefile]: Updated as you suggested.

File: scripts/build/build-chart-repo.sh [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  echo ">>> helm lint $chart"
+  helm lint "$chart"
+  chart_dest=$HELM_CHART_DEST/"`basename "$chart"`"
+  echo ">>> helm package -d $chart_dest $chart"
+  mkdir -p "$chart_dest"
+  helm package -d "$chart_dest" "$chart"
+done
+echo '>>>' "helm repo index --url https://$(dirname $GITHUB_PAGES_REPO).github.io/$(basename $GITHUB_PAGES_REPO) $HELM_CHART_DEST"
+helm repo index --url https://$(dirname $GITHUB_PAGES_REPO).github.io/$(basename $GITHUB_PAGES_REPO) $HELM_CHART_DEST
```
[the remark this answers is not in the document]
Comment (mkyc) [reply scripts/build/build-chart-repo.sh]: Done.

**Comment by Smiley73** [comment]
Please add the (C) header on all files. Kohl's is paying for your work, so we own it. ;-)

File: scripts/build/build-chart-repo.sh [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+set -euxo pipefail
+
+echo '>> Building charts...'
+find "$HELM_CHARTS_SOURCE" -mindepth 1 -maxdepth 1 -type d | while read -r chart; do
```
Comment (akavel) [inline scripts/build/build-chart-repo.sh]: After the merge of `prereqs/` and `operator/`, there will be only single directory left, right? So maybe we could then remove the `find` & `while`, and instead just set:

    chart="$HELM_CHARTS_SOURCE/eunomia-operator"

?

**Comment by mkyc** [comment]
@akavel thanks for comments, I'll introduce them as soon as @seanmalloy will confirm that I can merge those two charts  (done in https://github.com/KohlsTechnology/eunomia/pull/120/commits/242528b7a1ed2412101cf289ed2df0192f542c6c)

File: deploy/helm/eunomia-operator/Chart.yaml [hunk 11]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: eunomia-operator
-version: 1.0.0
+version: 3.2.1
```
[the remark this answers is not in the document]
Comment (seanmalloy) [reply deploy/helm/eunomia-operator/Chart.yaml]: I have the same question. What does `3.2.1` mean?

File: scripts/build/build-chart-repo.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (seanmalloy) [reply scripts/build/build-chart-repo.sh]: Yes ... please add the copyright header to all of the shell scripts.

File: DEVELOPMENT.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (seanmalloy) [reply DEVELOPMENT.md]: Yes ... let's go ahead and change this to `job/templates`.

**Comment by mkyc** [comment]
@akavel @vinny-sabatini @seanmalloy I believe I addressed all comments. Please check comments if you are ok with it and merge.

**Pull request merged**